src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs

```diff
@@ -102,5 +117,64 @@ public MyCodeAction(Func<CancellationToken, Task<Document>> createChangedDocumen
             {
             }
         }
+
+        private class FirstStatementTracker
+        {
+            private readonly Collection<StatementSyntax> _removed;
+            public ICollection<StatementSyntax> FirstStatements { get; }
+
+            public FirstStatementTracker()
+            {
+                _removed = new Collection<StatementSyntax>();
+                FirstStatements = new Collection<StatementSyntax>();
+            }
+
+            private bool WasFirst(StatementSyntax statementToRemove)
+            {
+                if (statementToRemove.IsFirstStatementInEnclosingBlock() || statementToRemove.IsFirstStatementInSwitchSection())
+                {
+                    return true;
+                }
+
+                bool wasMadeFirstByRemovalOfPredecessor = FirstStatements.Remove(statementToRemove);
+                if (wasMadeFirstByRemovalOfPredecessor)
+                {
+                    return true;
+                }
+
+                return false;
+            }
+
+            private StatementSyntax FindFirstUnremovedSuccessor(StatementSyntax statement)
+            {
+                var unremovedSuccessor = statement;
+                do
+                {
+                    unremovedSuccessor = unremovedSuccessor.GetNextStatement();
+                } while (_removed.Contains(unremovedSuccessor));
+
+                return unremovedSuccessor;
```

Comment:
can this ever be null?  if so, can we name this TryFindFirst...?  If it can't be null, can you add an assert?

Reply:
It is null if the statement is the only one in its parent, right?